driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorPublisherImpl.java

```diff
@@ -0,0 +1,160 @@
+import com.mongodb.reactivestreams.client.BatchCursorPublisher;
+import com.mongodb.reactivestreams.client.ClientSession;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static com.mongodb.reactivestreams.client.internal.PublisherCreator.sinkToCallback;
+
+public abstract class BatchCursorPublisherImpl<T> implements BatchCursorPublisher<T> {
+    private final ClientSession clientSession;
+    private final ReadConcern readConcern;
+    private final OperationExecutor executor;
+    private final ReadPreference readPreference;
+    private final boolean retryReads;
+    private Integer batchSize;
+
+    BatchCursorPublisherImpl(@Nullable final ClientSession clientSession, final OperationExecutor executor,
+                              final ReadConcern readConcern, final ReadPreference readPreference, final boolean retryReads) {
+        this.clientSession = clientSession;
+        this.executor = notNull("executor", executor);
+        this.readConcern = notNull("readConcern", readConcern);
+        this.readPreference = notNull("readPreference", readPreference);
+        this.retryReads = retryReads;
+    }
+
+    abstract AsyncReadOperation<AsyncBatchCursor<T>> asAsyncReadOperation();
+
+    @Nullable
+    public ClientSession getClientSession() {
+        return clientSession;
+    }
+
+    OperationExecutor getExecutor() {
+        return executor;
+    }
+
+    ReadPreference getReadPreference() {
+        return readPreference;
+    }
+
+    ReadConcern getReadConcern() {
+        return readConcern;
+    }
+
+    boolean getRetryReads() {
+        return retryReads;
+    }
+
+    @Nullable
+    public Integer getBatchSize() {
+        return batchSize;
+    }
+
+    @Override
+    public Publisher<T> batchSize(final int batchSize) {
+        this.batchSize = batchSize;
+        return this;
+    }
+
+    public Publisher<T> first() {
+        return batchCursor()
+                .flatMap(batchCursor -> Mono.create(sink -> {
+                    batchCursor.setBatchSize(1);
+                    Mono.from(batchCursor.next())
+                            .doOnTerminate(batchCursor::close)
+                            .doOnError(sink::error)
+                            .doOnSuccess(results -> {
+                                if (results == null || results.isEmpty()) {
+                                    sink.success();
+                                } else {
+                                    sink.success(results.get(0));
+                                }
+                            })
+                            .subscribe();
+                }));
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> subscriber) {
+        batchCursor()
```

Comment:
This is quite fun, we have to convert the `Mono<BatchCursor>` into a `Flux<T>` containing the cursors results.

Here we use the `flatMapMany` method to indicate we are producing a `Flux` from a `Mono` then we create a recursive `Flux` with a buffered overflow, that way internally we can just add each batch's item to the sink. The buffering will handle the demand from the subscriber.  The inProgress flag and request check limits the amount of data buffered from calls to `batchCursor.next`.